### PR TITLE
repart: Support minimizing verity partition hash always

### PR DESCRIPTION
**Pull request by mcassaniti**
This should hopefully address #27441 by creating the data partition as a loopback device when the verity hash partition is set to minimize. I'll add some tests after I get some feedback.

### CODE REVIEW

**Comment by DaanDeMeyer** [comment]
This will change existing behavior as the filesystem will not take up the full size anymore as specified in its partition. I don't think this approach is going to work out.

**Comment by yuwata** [comment]
@mcassaniti I am not sure about this works or not. But if you think this should work, then please add a test case about that.

**Comment by mcassaniti** [comment]
Here's some output showing it "works on my machine".

The TL;DR version is that we populate a file system with a mere 22MB of files but the FS is 2GB in size. I believe @DaanDeMeyer that this shows the FS will occupy whatever is set as `SizeMinBytes`.

This leaves the issue of growing the partition to occupy more of the disk. I get that the data partition in theory could be grown larger automatically (almost 10GB in the example below if the disk size was 10GB), but does that make a lot of sense for a verity protected partition that shouldn't be modified anyway? For A/B data partitions I'd much rather specify a maximum predictable size.

@yuwata I'll need some pointers on writing the test cases. I'm not even sure where to start.

```
root@rescue:/mnt/desktop/test-repart# cat 01-os.conf 

[Partition]
Type           = root
Format         = ext4

SizeMinBytes   = 2G
SizeMaxBytes   = 2G
ReadOnly=yes

CopyFiles      = /mnt/desktop/mkosi.installdir

VerityMatchKey = os
Verity         = data
root@rescue:/mnt/desktop/test-repart# cat 02-verity.conf 
[Partition]
Type           = root-verity

Minimize       = best

# Hash the data partition and populate with the hash contents
VerityMatchKey = os
Verity         = hash
root@rescue:/mnt/desktop/test-repart# systemd-repart --empty=create --size=auto --dry-run=no --definitions $PWD output.img
Pre-populating ext4 filesystem of partition 01-os.conf twice to calculate minimal verity partition size
/dev/loop0 successfully formatted as ext4 (label "root-x86-64", uuid 5ed1a72b-08b8-4bd1-8309-e752ee3e28bf)
Populating ext4 filesystem.
Successfully populated ext4 filesystem.
Minimal partition size of ext4 filesystem of partition 01-os.conf is 2.0G
/dev/loop0 successfully formatted as ext4 (label "root-x86-64", uuid 97acd7a3-aae3-4064-bac2-0c2c7dc44961)
Populating ext4 filesystem.
Successfully populated ext4 filesystem.
Pre-populating verity hash data of partition 02-verity.conf to calculate minimal partition size
Minimal partition size of verity hash partition 02-verity.conf is 136.5M
Automatically determined minimal disk image size as 2.1G.
Sized 'output.img' to 2.1G.
TYPE               LABEL              UUID                                 FILE           NODE            SIZE PADDING ROOTHASH                                                        
root-x86-64        root-x86-64        4128cc14-9ceb-73cd-c6e3-b4784a6a5969 01-os.conf     output.img1   → 2.0G    → 0B 4128cc149ceb73cdc6e3b4784a6a5969261fa575e38d5637d46d57ad5a1f0529
root-x86-64-verity root-x86-64-verity 261fa575-e38d-5637-d46d-57ad5a1f0529 02-verity.conf output.img2 → 136.5M    → 0B 4128cc149ceb73cdc6e3b4784a6a5969261fa575e38d5637d46d57ad5a1f0529
                                                                                                      Σ = 2.1G  Σ = 0B                                                                 

 ▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒░
 │                                                                                                                                                                                                  └─ 02-verity.c>
 └─ 01-os.conf                                                                                                                                                                                                    

Applying changes.
Wiped block device.
Discarded entire block device.
Successfully wiped file system signatures from future partition 0.
Successfully wiped file system signatures from future partition 1.
Copying in '/var/tmp/.#reparta74715cfdc1aad45' (2.0G) on block level into future partition 0.
Copying in of '/var/tmp/.#reparta74715cfdc1aad45' on block level completed.
Copying in '/var/tmp/.#repart65ba54c0af6876fd' (136.5M) on block level into future partition 1.
Copying in of '/var/tmp/.#repart65ba54c0af6876fd' on block level completed.
Adding new partition 0 to partition table.
Adding new partition 1 to partition table.
Writing new partition table.
All done.
TYPE               LABEL              UUID                                 FILE           NODE            SIZE PADDING ROOTHASH                                                        
root-x86-64        root-x86-64        4128cc14-9ceb-73cd-c6e3-b4784a6a5969 01-os.conf     output.img1   → 2.0G    → 0B 4128cc149ceb73cdc6e3b4784a6a5969261fa575e38d5637d46d57ad5a1f0529
root-x86-64-verity root-x86-64-verity 261fa575-e38d-5637-d46d-57ad5a1f0529 02-verity.conf output.img2 → 136.5M    → 0B 4128cc149ceb73cdc6e3b4784a6a5969261fa575e38d5637d46d57ad5a1f0529
                                                                                                      Σ = 2.1G  Σ = 0B                                                                 
root@rescue:/mnt/desktop/test-repart# losetup -P -f output.img
root@rescue:/mnt/desktop/test-repart# parted /dev/loop0 print
Model: Loopback device (loopback)
Disk /dev/loop0: 2292MB
Sector size (logical/physical): 512B/512B
Partition Table: gpt
Disk Flags: 

Number  Start   End     Size    File system  Name                Flags
 1      1049kB  2149MB  2147MB  ext4         root-x86-64
 2      2149MB  2292MB  143MB                root-x86-64-verity

root@rescue:/mnt/desktop/test-repart# mkdir tmp
root@rescue:/mnt/desktop/test-repart# mount /dev/loop0p1 tmp
root@rescue:/mnt/desktop/test-repart# df -hT "$PWD/tmp"
Filesystem     Type  Size  Used Avail Use% Mounted on
/dev/loop0p1   ext4  2.0G   22M  1.9G   2% /mnt/desktop/test-repart/tmp
root@rescue:/mnt/desktop/test-repart# 
```

**Comment by mcassaniti** [comment]
@yuwata The tests have been added. I'm not sure on the CI failures though.

**Comment by yuwata** [comment]
The CI failures are not related. Please ignore them.

**Comment by mcassaniti** [comment]
Can you suggest someone who can review please?

**Comment by poettering** [comment]
I am not sure I follow, i.e. what "minimization" is supposed to mean, but do note that for A/B update scenarios one usually wants to leave room for both the rootfs partition and its verity data to still grow. Hence it should be totally ok to build an image that reserves a large amount of space for verity without putting much in to it. And the verity partition size doesn't even have to be a direct function of the rootfs partition size that is shipped, since the rootfs might be the last partition on the golden image (which might be grown), and the verity for it could be earlier in the partition list but must be overcommited because it hence cannot be grown freely.

Anyway, I don't understand what this PR is supposed to do. Can you elaborate?

**Comment by mcassaniti** [comment]
It's actually for the case where an A/B style update is not used. I'm using this for a verified rescue image so my USB thumb drive can still boot properly under secure boot, get all the appropriate packages added and still maintain a reasonable level of security.

Now whether my rescue image is a good idea or not is debatable and I can handle criticism on that one, but the core idea is to get the entire image to be as small as possible and only updated by a complete blast over the top.
